include/seqan3/core/algorithm/concept_pre.hpp

```diff
@@ -0,0 +1,244 @@
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Resolves dependencies on the concepts.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+#include <type_traits>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3::detail
+{
+/*!\interface seqan3::detail::config_concept <>
+ * \ingroup core_algorithm
+ * \brief Concept for an algorithm configuration.
+ */
+/*!\name Requirements for seqan3::detail::config_concept
+ * \relates seqan3::detail::config_concept
+ * \brief You can expect these member types and the free function on all types that satisfy seqan3::detail::config_concept.
+ * \{
+ */
+
+/*!\fn      seqan3::detail::config_concept::data();
+ * \brief   Gives access to the stored configuration state.
+ * \returns The stored configuration state.
+ *
+ * \attention This is a concept requirement, not an actual function (however types satisfying this concept
+ * will provide an implementation).
+ */
+
+//!\}
+//!\cond
+template <typename config_t>
+concept bool config_concept = requires (config_t & cfg, config_t const & cfg_c)
+{
+    { cfg.data() };
+    { cfg_c.data() };
+    { std::move(cfg).data() };
+    { std::move(cfg_c).data() };
+
+    requires std::is_lvalue_reference_v<decltype(cfg.data())>;
+    requires std::is_lvalue_reference_v<decltype(cfg_c.data())> &&
+             std::is_const_v<std::remove_reference_t<decltype(cfg_c.data())>>;
+    requires std::is_rvalue_reference_v<decltype(std::move(cfg).data())>;
+    requires std::is_rvalue_reference_v<decltype(std::move(cfg_c).data())> &&
+             std::is_const_v<std::remove_reference_t<decltype(std::move(cfg_c).data())>>;
+};
```

Comment:
You are aware that this is satisfied by mayn opther types probably, too?

Reply:
After some cleanup I will have a more specific concept for this.
I needed a first feasibility draft that we can discuss before I invest to much time changing things afterwards.